perch/model.py

```diff
@@ -150,39 +168,90 @@ def get(cls, resource_id):
         if resource.get('type') != cls.resource_type:
             raise exceptions.NotFound()
 
+        if not include_deactivated and resource.get('state') == State.deactivated.name:
+            raise exceptions.NotFound()
+
         raise Return(cls(**resource))
 
     @classmethod
     @coroutine
-    def all(cls):
+    def all(cls, include_deactivated=False):
         """
         Get all resources
+        :param include_deactivated: Include deactivated resources in response
 
         :returns: list of Document instances
         :raises: SocketError, CouchException
         """
-        resources = yield cls.view.get(include_docs=True)
+        if include_deactivated:
+            resources = yield cls.view.get(include_docs=True)
+        else:
+            resources = yield cls.active_view.get(include_docs=True)
         raise Return([cls(**resource['doc']) for resource in resources['rows']])
 
     @classmethod
     @coroutine
     def create(cls, user, **kwargs):
+        resource = cls(**kwargs)
+
         if user:
             can_create = yield cls.can_create(user, **kwargs)
+            can_approve = yield resource.can_approve(user, **kwargs)
         else:
             can_create = False
+            can_approve = False
 
         if not can_create:
             err = 'User is not authorised to create this resource'
             raise exceptions.Unauthorized(err)
 
-        resource = cls(**kwargs)
+        # If user can approve resource and has not specified a state, approve on creation
+        if 'state' not in kwargs and can_approve:
+            resource._resource.update({'state': State.approved.name})
+
+        can_set_state = yield resource.validate_state_transition(user, None, resource.state, **kwargs)
+        if not can_set_state:
+            err = 'Cannot set initial state as {}'.format(resource.state.name)
+            raise exceptions.ValidationError(err)
+
         yield resource._save()
 
         raise Return(resource)
 
     @coroutine
+    def validate_state_transition(self, user, start_state, end_state, **kwargs):
+        """
+        Validate whether user can transition resource from start state to end state
+        :param user
+        :param start_state
+        :param end_state
+        :return: bool
+        """
+        if start_state == end_state:
+            raise Return(True)
+
+        allowed_transitions = self.state_transitions.get(start_state, [])
+
+        can_approve = yield self.can_approve(user, **kwargs)
+        if can_approve:
+            allowed_transitions += self.approval_state_transitions.get(start_state, [])
```

Comment:
This is mutating the class `state_transitions`, so the next time this method is executed the `approval_state_transition` will be in `state_transitions`, e.g.

```python
import perch

doc = perch.Document()
transitions = doc.state_transitions[perch.State.pending]
assert transitions == [perch.State.deactivated]

transitions += doc.approval_state_transitions[perch.State.pending]
assert transitions == [perch.State.deactivated, perch.State.approved, perch.State.rejected]

new_doc = perch.Document()
# this will raise an assertion error, the class now has [perch.State.deactivated, perch.State.approved, perch.State.rejected]
assert new_doc.state_transitions == [perch.State.deactivated]
```